Use static registration for execute flags

loganharbour opened the pull request:
Closes #20988

moosebuild commented:
Job [Documentation](https://civet.inl.gov/job/1039451/) on e537bf2 wanted to post the following:

View the site [here](https://mooseframework.inl.gov/docs/PRs/20989/site)

This comment will be updated on new commits.

loganharbour commented:
@lindsayad mind taking a look at this when you have a moment?

loganharbour commented on framework/src/base/Attributes.C:
```diff
@@ -28,9 +28,12 @@
 
 #include <algorithm>
 
+const ExecFlagType EXEC_ALL = defineExecFlag("ALL");
```
reminder: namespace this somewhere

lindsayad commented on framework/include/utils/GenericRegistry.h:
```diff
@@ -0,0 +1,175 @@
+{
+  if (id >= _id_to_item.size())
+    mooseError(_name, ": ID '", id, "' is not registered");
+  return _id_to_item[id];
```
I would return the dereferenced result of `find` which feels better to me considering the `const` method signature and your error check immediately above it. Moreover, it would also make me feel better about your call of this method from `item` where you only do a shared lock

loganharbour replied:
By deref-ed do you mean a copy? I think that's okay. For the PerfGraph, we do use a `PerfGraphSectionInfo` as the item (https://mooseframework.inl.gov/docs/doxygen/moose/classmoose_1_1internal_1_1PerfGraphSectionInfo.html), but... if this is only done statically, that should be fine?

lindsayad commented on framework/src/utils/PerfGraphRegistry.C:
```diff
@@ -51,83 +51,23 @@ PerfGraphRegistry::actuallyRegisterSection(const std::string & section_name,
 
-  return _id_to_section_info.size();
+  return registerItem(section_name, create);
+  ;
```
```suggestion
```

lindsayad commented on framework/src/base/ExecFlagRegistry.C:
```diff
@@ -0,0 +1,37 @@
+bool
+ExecFlagRegistry::isFlagRegistered(const MooseEnumItem & flag) const
+{
+  return keyExists(flag.name()) && item(id(flag.name())).id() == flag.id();
```
Is `item(id(flag.name())).id()` not guaranteed to equal `id(flag.name())`?

loganharbour replied:
Yeah - that's the whole point of the registry. Someone could go register `EXEC_FOO` with an ID (assigned by this) as 5. But then I could do randomly create a `MooseEnumItem` with name `EXEC_FOO` and id 6, in which case I wouldn't want _that_ item to be registered.

loganharbour replied:
Yeah - that's the whole point of the registry. Someone could go register `EXEC_FOO` with an ID (assigned by this) as 5. But then I could do randomly create a `MooseEnumItem` with name `EXEC_FOO` and id 6, in which case I wouldn't want _that_ item to be registered.

lindsayad replied:
Kk

lindsayad commented on framework/src/executors/Executor.C:
```diff
@@ -32,12 +34,12 @@ Executor::Executor(const InputParameters & parameters)
   if (!parameters.isParamSetByUser("begin_exec_flag"))
   {
-    _begin_flag = ExecFlagType("exec_" + _name + "_begin");
+    _begin_flag = moose::internal::getExecFlagRegistry().registerFlag("exec_" + _name + "_begin");
```
```suggestion
    _begin_flag = defineExecFlag("exec_" + _name + "_begin");
```

Am I using that right?

loganharbour replied:
Yep. I like this - expose fewer internals when possible

lindsayad commented on framework/src/executors/Executor.C:
```diff
@@ -32,12 +34,12 @@ Executor::Executor(const InputParameters & parameters)
   if (!parameters.isParamSetByUser("end_exec_flag"))
   {
-    _end_flag = ExecFlagType("exec_" + _name + "_end");
+    _end_flag = moose::internal::getExecFlagRegistry().registerFlag("exec_" + _name + "_end");
```
```suggestion
    _end_flag = defineExecFlag("exec_" + _name + "_end");
```

lindsayad reviewed (changes requested):
Overall looks very nice!

loganharbour committed:
Fix issue with pre and post aux

refs #20988

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/1137211/) on 57671b9 wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/20989/clang_format/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/20989/clang_format/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format eb0128f5d3aec72cce02c9a92e2c2b838d629325`

loganharbour commented on framework/include/utils/GenericRegistry.h:
```diff
@@ -0,0 +1,175 @@
+
+private:
+  /// Map of keys to IDs
+  std::unordered_map<Key, std::size_t> _key_to_id;
```
< 20 entries on average for the execution flags, ~100 on average for the perf graph. Of note here is that these are typically only accessed statically. Either on init, or on the first time a section is timed. Therefore, this really isn't accessed much at all. Unordered or not doesn't matter to me.

lindsayad replied:
In that case, fine to leave as-is I think

moosebuild commented:
Job [Coverage](https://civet.inl.gov/job/1137481/) on fa5da39 wanted to post the following:


## Framework coverage

<table>
  <thead>
    <tr>
      <th></th>
      <th>eb0128</th>
      <th colspan=4>#20989 fa5da3</th>
    </tr>
    <tr>
      <th></th>
      <th>Total</th>
      <th>Total</th>
      <th>+/-</th>
      <th>New</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td><b>Rate</b></td>
      <td>83.98%</td>
      <td>84.00%</td>
      <td>+0.02%</td>
      <td>100.00%</td>
    </tr>
    <tr>
      <td><b>Hits</b></td>
      <td>78658</td>
      <td>78679</td>
      <td>+21</td>
      <td>89</td>
    </tr>
    <tr>
      <td><b>Misses</b></td>
      <td>15004</td>
      <td>14985</td>
      <td>-19</td>
      <td>0</td>
    </tr>
  </tbody>
</table>

[Diff coverage report](https://mooseframework.inl.gov/docs/PRs/20989/coverage_diff/framework)

[Full coverage report](https://mooseframework.inl.gov/docs/PRs/20989/coverage/framework)


## Modules coverage

Coverage did not change

## Full coverage reports

<details>
<summary>Reports</summary>

- [`framework`](https://mooseframework.inl.gov/docs/PRs/20989/coverage/framework)
- [`chemical_reactions`](https://mooseframework.inl.gov/docs/PRs/20989/coverage/chemical_reactions)
- [`combined`](https://mooseframework.inl.gov/docs/PRs/20989/coverage/combined)
- [`contact`](https://mooseframework.inl.gov/docs/PRs/20989/coverage/contact)
- [`electromagnetics`](https://mooseframework.inl.gov/docs/PRs/20989/coverage/electromagnetics)
- [`external_petsc_solver`](https://mooseframework.inl.gov/docs/PRs/20989/coverage/external_petsc_solver)
- [`fluid_properties`](https://mooseframework.inl.gov/docs/PRs/20989/coverage/fluid_properties)
- [`fsi`](https://mooseframework.inl.gov/docs/PRs/20989/coverage/fsi)
- [`functional_expansion_tools`](https://mooseframework.inl.gov/docs/PRs/20989/coverage/functional_expansion_tools)
- [`geochemistry`](https://mooseframework.inl.gov/docs/PRs/20989/coverage/geochemistry)
- [`heat_conduction`](https://mooseframework.inl.gov/docs/PRs/20989/coverage/heat_conduction)
- [`level_set`](https://mooseframework.inl.gov/docs/PRs/20989/coverage/level_set)
- [`misc`](https://mooseframework.inl.gov/docs/PRs/20989/coverage/misc)
- [`navier_stokes`](https://mooseframework.inl.gov/docs/PRs/20989/coverage/navier_stokes)
- [`peridynamics`](https://mooseframework.inl.gov/docs/PRs/20989/coverage/peridynamics)
- [`phase_field`](https://mooseframework.inl.gov/docs/PRs/20989/coverage/phase_field)
- [`porous_flow`](https://mooseframework.inl.gov/docs/PRs/20989/coverage/porous_flow)
- [`ray_tracing`](https://mooseframework.inl.gov/docs/PRs/20989/coverage/ray_tracing)
- [`rdg`](https://mooseframework.inl.gov/docs/PRs/20989/coverage/rdg)
- [`reactor`](https://mooseframework.inl.gov/docs/PRs/20989/coverage/reactor)
- [`richards`](https://mooseframework.inl.gov/docs/PRs/20989/coverage/richards)
- [`stochastic_tools`](https://mooseframework.inl.gov/docs/PRs/20989/coverage/stochastic_tools)
- [`tensor_mechanics`](https://mooseframework.inl.gov/docs/PRs/20989/coverage/tensor_mechanics)
- [`thermal_hydraulics`](https://mooseframework.inl.gov/docs/PRs/20989/coverage/thermal_hydraulics)
- [`xfem`](https://mooseframework.inl.gov/docs/PRs/20989/coverage/xfem)

</details>



This comment will be updated on new commits.

loganharbour committed:
Use static registration for execute flags

refs #20988

- Removes the need to register execute flags from within an application,
instead a static registry across all applications is used
- Removes the need to const cast away static vars to change IDs
for exec flags under the hood

lindsayad commented:
So you blew up the `GenericRegistry`?

loganharbour committed:
Remove unused parameters

refs #20988

loganharbour commented:
Yeah - sorry for the yuge change. Thought about this more and realized that it wasn't really necessary anymore to both define static flags and then register them in each app. The registry now just holds an enum of all of the flags, and said multi enum is used for all of the tracking. This makes it pretty straightforward

loganharbour committed:
Add missing header and clean up locks

refs #20988

loganharbour committed:
Correct duplicate registration check

refs #20988

loganharbour committed:
Add test for duplicate exec flag registration error checking

refs #20988

loganharbour committed:
Correct logic for enum find via enum

refs #20988

The enum should only be found if the ID and name are the same

loganharbour committed:
defineExecFlag -> registerExecFlag

refs #20988

GiudGiud commented on test/src/userobjects/ExecFlagRegistryErrorTest.C:
```diff
@@ -4,7 +4,7 @@
 //* https://github.com/idaholab/moose/blob/master/COPYRIGHT
 //*
-//* Licensed under LGPL 2.1, please see LICENSE for details
+//* Licensed under LGPL 2.1, please see LICENSE for registerExecFlag
```
?

lindsayad replied:
Wow that's a fancy license!

loganharbour replied:
...still learning vscode find and replace

GiudGiud commented on framework/include/base/Attributes.h:
```diff
@@ -129,17 +130,21 @@ class AttribVectorTags : public AttribTagBase
 class AttribExecOns : public Attribute
 {
 public:
-  typedef unsigned int Key;
```
what s up with unsigned int -> int

loganharbour replied:
`unsigned int` was previously wrong. This represents a flag value in a `MultiMooseEnum`, in which the IDs are `int` (we've supported negative values for a long time)

GiudGiud commented on test/src/base/MooseTestApp.C:
```diff
@@ -98,10 +98,7 @@ MooseTestApp::registerAll(Factory & f, ActionFactory & af, Syntax & s, bool use_
-    auto & factory = f; // for resiterSyntax macros
-
-    registerExecFlag(EXEC_JUST_GO);
+    auto & syntax = s; // for resiterSyntax macros
```
```suggestion
    auto & syntax = s; // for registerSyntax macros
```

loganharbour replied:
Boo this isn't even my line

loganharbour committed:
Update documentation for new exec flag registration

refs #20988

loganharbour committed:
Remove now unnecessary exec flag script

refs #20988

loganharbour commented:
Good to go, patches are ready

lindsayad commented on framework/include/base/Attributes.h:
```diff
@@ -148,7 +153,7 @@ class AttribExecOns : public Attribute
 
 private:
-  std::vector<unsigned int> _vals;
+  std::vector<int> _vals;
```
```suggestion
  std::vector<Key> _vals;
```

lindsayad commented on framework/include/warehouses/ExecuteMooseObjectWarehouse.h:
```diff
@@ -205,32 +202,14 @@ void
-        _execute_objects[*iter].addObject(object, tid);
-    }
-  }
+  if (const std::shared_ptr<SetupInterface> ptr = std::dynamic_pointer_cast<SetupInterface>(object))
```
```suggestion
  if (const auto ptr = std::dynamic_pointer_cast<SetupInterface>(object))
```

lindsayad commented on framework/src/base/Moose.C:
```diff
@@ -383,27 +378,6 @@ registerActions(Syntax & syntax,
   registerTask("finish_input_file_output", false);
 }
 
-void
-registerExecFlags(Factory & factory)
-{
-  registerExecFlag(EXEC_NONE);
-  registerExecFlag(EXEC_INITIAL);
-  registerExecFlag(EXEC_LINEAR);
-  registerExecFlag(EXEC_NONLINEAR);
-  registerExecFlag(EXEC_TIMESTEP_END);
-  registerExecFlag(EXEC_TIMESTEP_BEGIN);
-  registerExecFlag(EXEC_FINAL);
-  registerExecFlag(EXEC_FORCED);
-  registerExecFlag(EXEC_FAILED);
-  registerExecFlag(EXEC_CUSTOM);
-  registerExecFlag(EXEC_TRANSFER);
-  registerExecFlag(EXEC_SUBDOMAIN);
-  registerExecFlag(EXEC_PRE_DISPLACE);
-  registerExecFlag(EXEC_SAME_AS_MULTIAPP);
-  registerExecFlag(EXEC_PRE_MULTIAPP_SETUP);
-  registerExecFlag(EXEC_PRE_KERNELS);
-}
-
```
nice to see this gone

loganharbour committed:
Use Key globally in AttribExecOns

refs #20988

loganharbour committed:
Auto all the things!

refs #20988
